### PR TITLE
Add mypyc output

### LINKED ISSUE #83: add output for "mypy and a compiler that works with mypyc"
### Comment:

It would be _lovely_ if there was a `mypyc` output which also dragged along an appropriate compiler, such that other packages could depend on it.

it seems like `mypyc` wouldn't be an _awful_ name for it, but maybe not _strictly_ accurate, as it _might_ theoretically work otherwise with a compiler-of-opportunity.

Motivation: https://github.com/conda-forge/staged-recipes/pull/21071

### PR DESCRIPTION
**Pull request by bollwyvl**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

Notes:

- closes #83

### CODE REVIEW

File: recipe/meta.yaml [hunk 1]
```diff
@@ -1,75 +1,139 @@
...
+        - m2w64-toolchain  # [win]
+    test:
+      files:
+        - test_mypyc.py
+      requires:
+        - pip
+      commands:
+        - pip check
+        - mypyc --help
+        - python test_mypyc.py  # [not(aarch64 or ppc64le)]
```
Comment (bollwyvl) [inline recipe/meta.yaml:108]: not proud of this, but didn't know what else to do

**Comment by bollwyvl** [comment]
@conda-forge/mypy ready for review

File: recipe/meta.yaml (hunk 1 above)
Comment (ocefpaf) [reply recipe/meta.yaml:108]: It is more common than we would like to admit. No judgments here ;-p

**Comment by ocefpaf** [comment]
@bollwyvl you know the rule, you break it, you bought it. So why not add yourself as a maintainer too ;-p

File: recipe/meta.yaml (hunk 1 above)
Comment (bollwyvl) [reply recipe/meta.yaml:108]: here's [a fail](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=609636&view=logs&j=bb1c2637-64c6-57bd-9ea6-93823b2df951&t=350df31b-3291-5209-0bb7-031395f0baa1&l=7528) if it jostles anything loose in a reviewer's mind

```
running build_ext
building 'fib' extension
creating build/temp.linux-aarch64-cpython-310
creating build/temp.linux-aarch64-cpython-310/build
$PREFIX/bin/x86_64-conda-linux-gnu-cc -Wno-unused-result -Wsign-compare -DNDEBUG -fwrapv -O3 -Wall -fPIC -O3 -isystem $PREFIX/include -fPIC -O3 -isystem $PREFIX/include -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem $PREFIX/include -fdebug-prefix-map=$SRC_DIR=/usr/local/src/conda/mypyc-0.990 -fdebug-prefix-map=$PREFIX=/usr/local/src/conda-prefix -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem $PREFIX/include -fPIC -I$PREFIX/lib/python3.10/site-packages/mypyc/lib-rt -I$PREFIX/include/python3.10 -c build/__native.c -o build/temp.linux-aarch64-cpython-310/build/__native.o -O3 -g1 -Werror -Wno-unused-function -Wno-unused-label -Wno-unreachable-code -Wno-unused-variable -Wno-unused-command-line-argument -Wno-unknown-warning-option -Wno-unused-but-set-variable
creating build/lib.linux-aarch64-cpython-310
$PREFIX/bin/x86_64-conda-linux-gnu-cc -shared -Wl,--allow-shlib-undefined -Wl,-rpath,$PREFIX/lib -Wl,-rpath-link,$PREFIX/lib -L$PREFIX/lib -Wl,--allow-shlib-undefined -Wl,-rpath,$PREFIX/lib -Wl,-rpath-link,$PREFIX/lib -L$PREFIX/lib -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,--allow-shlib-undefined -Wl,-rpath,$PREFIX/lib -Wl,-rpath-link,$PREFIX/lib -L$PREFIX/lib -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem $PREFIX/include -fdebug-prefix-map=$SRC_DIR=/usr/local/src/conda/mypyc-0.990 -fdebug-prefix-map=$PREFIX=/usr/local/src/conda-prefix -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem $PREFIX/include build/temp.linux-aarch64-cpython-310/build/__native.o -o build/lib.linux-aarch64-cpython-310/fib.cpython-310-aarch64-linux-gnu.so
/home/conda/feedstock_root/build_artifacts/mypy-split_1668016285943/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/bin/../lib/gcc/x86_64-conda-linux-gnu/12.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: skipping incompatible /home/conda/feedstock_root/build_artifacts/mypy-split_1668016285943/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/libgcc_s.so.1 when searching for libgcc_s.so.1
/home/conda/feedstock_root/build_artifacts/mypy-split_1668016285943/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/bin/../lib/gcc/x86_64-conda-linux-gnu/12.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: skipping incompatible /home/conda/feedstock_root/build_artifacts/mypy-split_1668016285943/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/libgcc_s.so.1 when searching for libgcc_s.so.1
/home/conda/feedstock_root/build_artifacts/mypy-split_1668016285943/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/bin/../lib/gcc/x86_64-conda-linux-gnu/12.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: skipping incompatible /home/conda/feedstock_root/build_artifacts/mypy-split_1668016285943/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/libgcc_s.so.1 when searching for libgcc_s.so.1
/home/conda/feedstock_root/build_artifacts/mypy-split_1668016285943/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/bin/../lib/gcc/x86_64-conda-linux-gnu/12.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: skipping incompatible /home/conda/feedstock_root/build_artifacts/mypy-split_1668016285943/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/libgcc_s.so.1 when searching for libgcc_s.so.1
/home/conda/feedstock_root/build_artifacts/mypy-split_1668016285943/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/bin/../lib/gcc/x86_64-conda-linux-gnu/12.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: skipping incompatible /home/conda/feedstock_root/build_artifacts/mypy-split_1668016285943/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/libgcc_s.so.1 when searching for libgcc_s.so.1
/home/conda/feedstock_root/build_artifacts/mypy-split_1668016285943/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/bin/../lib/gcc/x86_64-conda-linux-gnu/12.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: skipping incompatible /home/conda/feedstock_root/build_artifacts/mypy-split_1668016285943/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/libgcc_s.so.1 when searching for libgcc_s.so.1
/home/conda/feedstock_root/build_artifacts/mypy-split_1668016285943/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/bin/../lib/gcc/x86_64-conda-linux-gnu/12.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: skipping incompatible /home/conda/feedstock_root/build_artifacts/mypy-split_1668016285943/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/libgcc_s.so.1 when searching for libgcc_s.so.1
/home/conda/feedstock_root/build_artifacts/mypy-split_1668016285943/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/bin/../lib/gcc/x86_64-conda-linux-gnu/12.2.0/../../../../x86_64-conda-linux-gnu/bin/ld: skipping incompatible /home/conda/feedstock_root/build_artifacts/mypy-split_1668016285943/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/libgcc_s.so.1 when searching for libgcc_s.so.1
copying build/lib.linux-aarch64-cpython-310/fib.cpython-310-aarch64-linux-gnu.so -> 
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ImportError: /home/conda/feedstock_root/build_artifacts/mypy-split_1668016285943/test_tmp/fib.cpython-310-aarch64-linux-gnu.so: cannot open shared object file: No such file or directory
Traceback (most recent call last):
  File "/home/conda/feedstock_root/build_artifacts/mypy-split_1668016285943/test_tmp/test_mypyc.py", line 23, in <module>
    subprocess.check_call([sys.executable, "-c", "import fib"])
  File "/home/conda/feedstock_root/build_artifacts/mypy-split_1668016285943/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/lib/python3.10/subprocess.py", line 369, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/home/conda/feedstock_root/build_artifacts/mypy-split_1668016285943/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/bin/python', '-c', 'import fib']' returned non-zero exit status 1.
```

**Comment by bollwyvl** [comment]
> add yourself as a maintainer too

aw, man, ain't that the way, just wanted to do some demos for a talk, end up maintaining software...

**Comment by bollwyvl** [comment]
@conda-forge-admin please rerender

File: recipe/meta.yaml [hunk 2]
```diff
@@ -1,75 +1,139 @@
...
+        - ls $SP_DIR/mypy/*.so  # [linux and not (aarch64 and py==310)]
+
+  - name: mypyc
+    requirements:
+      host:
+        - python
+      run:
+        - python
+        - {{ pin_subpackage("mypy", exact=True) }}
+        - gcc_linux-64     # [linux]
```
Comment (bollwyvl) [inline recipe/meta.yaml:97]: or maybe the suggestion of using the `compilers` package? how much does that bring along?

Comment (bollwyvl) [reply recipe/meta.yaml]: ah, `cxx` and `fortran`... while very cool, `mypyc` wouldn't know what to do with those... would just `c-compiler` work?

**Comment by bollwyvl** [comment]
@conda-forge-admin please rerender

File: recipe/meta.yaml (hunk 2 above)
Comment (bollwyvl) [reply recipe/meta.yaml]: tyring `c-compiler` on https://github.com/conda-forge/mypyc-ipython-feedstock/pull/2

Comment (bollwyvl) [reply recipe/meta.yaml]: looks good over there, will update here to use `c-compiler`

**Comment by bollwyvl** [comment]
@conda-forge-admin please rerender

File: recipe/meta.yaml (hunk 2 above)
Comment (ocefpaf) [reply recipe/meta.yaml]: Hey, I did not know we had a c-compiler package! Nice!

Comment (bollwyvl) [reply recipe/meta.yaml]: Hmm:

does `mypyc` need a `{{ compiler('c') }}` in the `build` so it syncs up the versions correctly, or will the dep on `mypy` take care of that?

```
{
  "mypy-0.990-py38hdacc76a_2": {
    "recipe": {
      "build_platform": "linux-64",
      "c_compiler": "gcc",
      "c_compiler_version": "10",
      "channel_targets": "conda-forge main",
      "python": "3.8.* *_cpython",
      "target_platform": "linux-aarch64"
    }
  },
  "mypyc-0.990-py38ha081722_2": {
    "recipe": {
      "c_compiler": "gcc",
      "channel_targets": "conda-forge main",
      "python": "3.8.* *_cpython",
      "target_platform": "linux-aarch64"
    }
  }
}
```

Comment (bollwyvl) [reply recipe/meta.yaml]: Great, all :heavy_check_mark: : other than this question about `build` deps to get the right version of the compiler, and there's not much else to do here.i

Comment (ocefpaf) [reply recipe/meta.yaml]: > does mypyc need a {{ compiler('c') }} in the build so it syncs up the versions correctly, or will the dep on mypy take care of that?

In theory I think we are OK. If compilers is updated the c-compiled will be update too and things should sync up.

File: recipe/meta.yaml (hunk 1 above)
Comment (bollwyvl) [reply recipe/meta.yaml]: I think we have this sorted with `c-compiler`!

File: recipe/meta.yaml (hunk 2 above)
Comment (bollwyvl) [reply recipe/meta.yaml]: Great. I'm willing to wait and see. At present `mypyc` is still kinda niche, and if you used it in a build, you'd still need to add `{{ compiler('c') }}` presumably, so it's really just the interactive case that would be "exciting". Think we're good here!

**Comment by bollwyvl** [comment]
I think everything has be addressed :tada: !

**Comment by bollwyvl** [comment]
thanks!

File: recipe/meta.yaml [hunk 3]
```diff
@@ -1,75 +1,137 @@
...
+        - dir %SP_DIR%\mypy\*.pyd /s /b  # [win]
+        - ls $SP_DIR/mypy/*.so  # [linux and not (aarch64 and py==310)]
+
+  - name: mypyc
+    requirements:
+      host:
+        - python
+      run:
+        - {{ pin_subpackage("mypy", exact=True) }}
+        - c-compiler
```
Comment (jakirkham) [inline recipe/meta.yaml:96]: Think we want this

```suggestion
        - {{ compiler("c") }}
```

**Review by jakirkham — COMMENTED** [review]
Sorry for being late here. Had one comment below.

Another question is whether we want `mypyc` to have [`compiler` syntax like other compilers]( https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/261390f932c60365a0037a1bd41ffb39c4a8118e/recipe/conda_build_config.yaml#L7-L10 )

File: recipe/meta.yaml (hunk 3 above)
Comment (bollwyvl) [reply recipe/meta.yaml:96]: ah, had been working off https://github.com/conda-forge/mypy-feedstock/issues/83#issuecomment-1308711145. will happily defer to whatever's best, especially if we find a case where it's busted... have only been able to get test reports locally on linux, despite cajoling folk with windows/apple boxen...

Comment (bollwyvl) [reply recipe/meta.yaml:96]: related: https://github.com/conda-forge/staged-recipes/pull/21144

Comment (jakirkham) [reply recipe/meta.yaml:96]: Submitted PR ( https://github.com/conda-forge/mypy-feedstock/pull/87 )

**Pull request merged**